DeNTAS  :De Novo Transcriptome Analysis & Statistics

This file contains a basic description of DeNTAS' analysis pipeline, details of the softwares requirements
and user instructions to run the program from a local machine. Please see the DeNTAS's documentation for further
information <link here?> 

DeNTAS is a software tool for the statistical analysis and visualisation of Transcriptome datasets
generated by de novo assemblies of RNA-seq datasets.

User input:
1) Raw assembled transcripts in FASTA format with FPKM values
2) Organism that samples were derived from
3) Experimental groups from the samples

Analysis:
1) Identification of transcripts via local blast on Apocrita (QMUL's high performace computing cluster)
2) Differential gene expression analysis and statistics conducted in R using the limma package

Results returned:
1) Unsupervised data exploration: Principla component and hierachical clustering (dendrograms) plots
   both for the full set of genes and for the subset of genes determined to be significantly differntially
   expressed
2) Volcano plots depicting each experimental comparrison. Those genes that have a logFoldChange > 2 and an
   adjusted p-value < 0.05 are highleted in red and labelled
3) Heatmap depicting the top 100 differntially regulated genes
4) Interactive table listing all differential regulated genes, by refseq ID, gene symbol, full gene name 
   and showing their e-value for each experimental comparison 


DeNTAS harnesses the power of high performance cluster computing to increase the speed of analysis.
Therefore execution of the software on a local machine has the following requirements the user must:
1) Have secure access to Apocrita without the need for a pasword 
   >> ssh-keygen (hit enter 3 times)
   >> ssh-copy-id <username>@login.hpc.qmul.ac.uk (input your apocrita password)
   >> ssh <username>@login.hpc.qmul.ac.uk (test connectivity)

2) Make changes to various filenames in the app to reflect your apocrita username:
   a) change 
